File: python/2022/01/main.py
#!/usr/bin/env python3

import sys
sys.path.append("../../")
sys.path.append("../")
import util as u

###############
debug = False
###############

file = "sample.txt" if debug else "input.txt"

def read_input():
    return u.read_lines(file)

def part1():
    A = read_input()
    k = 0
    tmp = 0
    for a in A:
        if len(a) == 0:
            if tmp > k:
                k = tmp
            tmp = 0
        else:
            tmp += int(a)
    if tmp > k:
        k = tmp
    print("Part 1:",k)

def part2():
    A = read_input()
    k = 0
    tmp = 0
    n = []
    for a in A:
        if len(a) == 0:
            n.append(tmp)
            tmp = 0
        else:
            tmp += int(a)
    n.append(tmp)
    n = sorted(n)
    k = n[-1] + n[-2] + n[-3]
    print("Part 2:",k)

part1()
part2()

File: pascal/2022/01/main.pas
program main;
uses 
	util, sysutils;

type
	m_arr = array [0..2] of longint;

const
	debug = false;

	sampleFile = 'sample.txt';
	inputFile = 'input.txt';	

var
	fp: string;
	f: text;
	
	procedure part1;
	var 
		A: array of string;
		i, tmp, max: longint;	
		aa: string;
	begin		
		A := util.readInput(f);
		i := 0;
		tmp := 0;
		max := -1;
		for aa in A do begin
			if aa = '' then begin
				if tmp > max then begin
					max := tmp;
				end;
				tmp := 0;
			end else begin
				tmp += StrToInt(aa);
			end
		end;
		if tmp > max then begin
			i := tmp;
		end else begin
			i := max;
		end;
		WriteLn('Part 1: ',i);
	end;

	procedure add(v: longint; var max: m_arr);
	begin
(*		writeln('>',v,' ',max[0],' ',max[1],' ',max[2]);*)
		if v > max[0] then begin
			max[2] := max[1];
			max[1] := max[0];
			max[0] := v;
		end else if v > max[1] then begin
			max[2] := max[1];
			max[1] := v;
		end else if v > max[2] then begin
			max[2] := v;
		end;
	end;

	procedure part2;
	var 
		A: array of string;
		i, tmp: longint;	
		max: array [0..2] of longint = (0,0,0);
		aa: string;
	begin		
		A := util.readInput(f);
		i := 0;
		tmp := 0;
		for aa in A do begin
			if aa = '' then begin
				add(tmp, max);
				tmp := 0;
			end else begin
				tmp += StrToInt(aa);
			end
		end;
		add(tmp, max);
		i := max[0]+max[1]+max[2];
		WriteLn('Part 2: ',i);
	end;
begin
	if debug then
		fp := sampleFile
	else
		fp := inputFile;	
	assign(f, fp);

	part1;
	part2;

end.
